src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -780,62 +813,89 @@ internal static bool Dispatch()
-                        }
+                        workQueue.EnsureThreadRequested();
+                    }
 
-                        //
-                        // No work.
-                        // If we missed a steal, though, there may be more work in the queue.
-                        // Instead of looping around and trying again, we'll just request another thread.  Hopefully the thread
-                        // that owns the contended work-stealing queue will pick up its own workitems in the meantime,
-                        // which will be more efficient than this thread doing it anyway.
-                        //
-                        if (missedSteal)
-                        {
-                            workQueue.EnsureThreadRequested();
-                        }
+                    return true;
+                }
+            }
 
-                        // Tell the VM we're returning normally, not because Hill Climbing asked us to return.
-                        return true;
-                    }
+            {
+                object? secondWorkItem = DequeueWithPriorityAlternation(workQueue, tl, out bool missedSteal);
+                if (secondWorkItem != null && Interlocked.CompareExchange(ref _nextWorkItemToProcess, secondWorkItem, null) != null)
+                {
+                    tl.workStealingQueue.LocalPush(secondWorkItem);
                 }
+                if (secondWorkItem != null || missedSteal)
+                {
+                    // TODO: Update comment below
+
+                    // A work item was successfully dequeued, and there may be more work items to process. Request a thread to
+                    // parallelize processing of work items, before processing more work items. Following this, it is the
+                    // responsibility of the new thread and other enqueuers to request more threads as necessary. The
+                    // parallelization may be necessary here for correctness (aside from perf) if the work item blocks for some
+                    // reason that may have a dependency on other queued work items.
 
-                // A work item was successfully dequeued, and there may be more work items to process. Request a thread to
-                // parallelize processing of work items, before processing more work items. Following this, it is the
-                // responsibility of the new thread and other enqueuers to request more threads as necessary. The
-                // parallelization may be necessary here for correctness (aside from perf) if the work item blocks for some
-                // reason that may have a dependency on other queued work items.
-                workQueue.EnsureThreadRequested();
+                    workQueue._separated.queueProcessingStage = (int)QueueProcessingStage.Scheduled;
+                    // I thought we might want to add a MemoryBarrier here but not sure about it
```

Comment:
Don't need a barrier here. Requesting a thread is a strongly-ordered operation, the thread that would be released would see the updated state.